<launch>



    <!-- mono or stereo and what ros bag to play -->
    <arg name="max_cameras" default="1" />
    <arg name="use_stereo"  default="false" />
    <arg name="bag_start"   default="5" />
    <arg name="bag"         default="/home/eh420/datasets/lab/lab_depth.bag" />
    <!--arg name="bag"         default="/home/eh420/datasets/vins_rgbd/handheld_Normal.bag" /-->
    <!--arg name="bag"         default="/home/eh420/datasets/vins_rgbd/Simple.bag" /-->

    <!-- imu starting thresholds -->
    <arg name="init_window_time"  default="1.0" />
    <arg name="init_imu_thresh"   default="0.56" />

    <!-- saving trajectory path and timing information -->
    <arg name="dosave"      default="false" />
    <arg name="dotime"      default="false" />
    <arg name="path_est"    default="/tmp/traj_estimate.txt" />
    <arg name="path_time"   default="/tmp/traj_timing.txt" />


    <!-- MASTER NODE! -->
    <node name="run_subscribe_msckf_rgbd" pkg="ov_msckf" type="run_subscribe_msckf_rgbd" output="screen" clear_params="true" required="true">

        <!-- bag topics -->
        <param name="topic_imu"      type="string" value="/camera/imu" />
        <param name="topic_camera0"  type="string" value="/camera/color/image_raw" />
        <!--param name="topic_depth_camera"  type="string" value="/camera/aligned_depth_to_color/image_raw" /-->
	<param name="topic_depth_camera"  type="string" value="/camera/depth/image_rect_raw" />

        <!-- world/filter parameters -->
        <param name="use_fej"                type="bool"   value="true" />
        <param name="use_imuavg"             type="bool"   value="true" />
        <param name="use_rk4int"             type="bool"   value="true" />
        <param name="use_stereo"             type="bool"   value="$(arg use_stereo)" />
        <param name="calib_cam_extrinsics"   type="bool"   value="true" />
        <param name="calib_cam_intrinsics"   type="bool"   value="true" />
        <param name="calib_cam_timeoffset"   type="bool"   value="true" />
        <param name="calib_camimu_dt"        type="double" value="-0.024291604791660072" />
        <param name="max_clones"             type="int"    value="20" />
        <param name="max_slam"               type="int"    value="50" />
        <param name="max_slam_in_update"     type="int"    value="25" /> <!-- 25 seems to work well -->
        <param name="max_msckf_in_update"    type="int"    value="999" />
        <param name="max_cameras"            type="int"    value="$(arg max_cameras)" />
        <param name="dt_slam_delay"          type="double" value="1" />
        <param name="init_window_time"       type="double" value="$(arg init_window_time)" />
        <param name="init_imu_thresh"        type="double" value="$(arg init_imu_thresh)" />
        <rosparam param="gravity">[0.0, 0.0, 9.805]</rosparam>
        <param name="feat_rep_msckf"         type="string" value="GLOBAL_3D" />
        <param name="feat_rep_slam"          type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />
        <param name="feat_rep_aruco"         type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />

        <!-- timing statistics recording -->
        <param name="record_timing_information"   type="bool"   value="$(arg dotime)" />
        <param name="record_timing_filepath"      type="string" value="$(arg path_time)" />

        <!-- tracker/extractor properties -->
        <param name="use_klt"          type="bool"   value="true" />
        <param name="num_pts"          type="int"    value="200" />
        <param name="fast_threshold"   type="int"    value="15" />
        <param name="grid_x"           type="int"    value="5" />
        <param name="grid_y"           type="int"    value="5" />
        <param name="min_px_dist"      type="int"    value="5" />
        <param name="knn_ratio"        type="double" value="0.65" />

        <!-- aruco tag/mapping properties -->
        <param name="use_aruco"        type="bool"   value="false" />
        <param name="num_aruco"        type="int"    value="1024" />
        <param name="downsize_aruco"   type="bool"   value="true" />

        <!-- sensor noise values / update -->
        <param name="up_msckf_sigma_px"            type="double"   value="1" />
        <param name="up_msckf_chi2_multipler"      type="double"   value="1" />
        <param name="up_slam_sigma_px"             type="double"   value="1" />
        <param name="up_slam_chi2_multipler"       type="double"   value="1" />
        <param name="up_aruco_sigma_px"            type="double"   value="1" />
        <param name="up_aruco_chi2_multipler"      type="double"   value="1" />
        <param name="gyroscope_noise_density"      type="double"   value="0.00262" />
        <param name="gyroscope_random_walk"        type="double"   value="3e-06" />
        <param name="accelerometer_noise_density"  type="double"   value="0.02568" />
        <param name="accelerometer_random_walk"    type="double"   value="0.000209" />

        <!-- camera intrinsics -->
        <rosparam param="cam0_wh">[640, 480]</rosparam>
        <param name="cam0_is_fisheye" type="bool" value="false" />
        <rosparam param="cam0_k">[1070.28079054, 1075.23462271, 360.12392485, 113.88568488]</rosparam>
        <rosparam param="cam0_d">[0.17264227, 1.54118064, -15.65262936, 18.82536235]</rosparam>


        <!-- camera extrinsics -->
        <rosparam param="T_C0toI">
            [
            0.99976483, -0.00322587,  0.0214448,  -0.01365905,
            0.00483444,  0.99714267, -0.07538645, -0.05238932,
            -0.02114034,  0.07547239,  0.99692377, -0.26117454,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>


    </node>


    <!-- play the dataset -->
    <node pkg="rosbag" type="play" name="rosbag" args="-u 10 -r 0.5 -d 1 -s $(arg bag_start) $(arg bag)" required="true"/>

    <node pkg="tf"
        type="static_transform_publisher"
        name="world_to_global"
        args="0.0 0.0 0.0 0 0 1.57 world global 100" />    




    <!-- record the trajectory if enabled -->
    <group if="$(arg dosave)">
        <node name="recorder_estimate" pkg="ov_eval" type="pose_to_file" output="screen">
            <param name="topic"      type="str" value="/ov_msckf/poseimu" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_est)" />
        </node>
    </group>

</launch>
